### PR TITLE
[WIP] WebAuthn implementation

### LINKED ISSUE #6892: Implement WebAuthn support
WebAuthn gained some traction in recent months. Specifically now that most android devices are certified security keys, webauthn becomes very appealing!

What needs  to happen for this is:
 - a new setting to allow the user to login with **only** a username (no password!)
 - implement webauthn to run in that flow

As Gitea is written in Go, you're golden when it comes to WebAuthn support. There is a quite well documented library: https://github.com/duo-labs/webauthn

Now you need a biometric device to login (for example to https://webauthn.io/). It's easy on android, you just place your finger on the fingerprint sensor and you're in. On the desktop this is a bit more tricky. As this is all still quite new, i guess more applications will pop up that allow you to use your phone's biometric sensors to login on desktop sites.. But one way to go about this now is by using the Krypton authenticator app. It allows you to do exactly that.

So, can we have this please? :)

Best regards,
Mark

### PR DESCRIPTION
**Pull request by e3b0c442**
Fixes #6892.

This PR is tracking my progress on updating the legacy U2F security key implementation to use standard WebAuthn, as discussed in #6892. I'm opening this as a draft PR so that my (slow) progress can be tracked and discussed in real time. If you have any feedback at all as I progress on this, I welcome it! I will convert the PR to a standard PR once all of the pieces are in place.

This change will replace the https://github.com/tstranex/u2f module with ~~https://github.com/duo-labs/webauthn~~ https://github.com/e3b0c442/warp on the backend with the necessary model adjustments and remove the Google u2fApi javascript library in favor of the WebAuthn standard API that is now universally available across current browsers. There will be no changes to the registration or login flows at this time -- that is to say, I'm only implementing the familiar 2nd-factor auth flow, and not the single-factor passwordless flow. I'm choosing to make this a clean break due to the fact that 1) All browsers that supported U2F now support WebAuthn and have for some time, and 2) while existing U2F credentials can be used for WebAuthn, credentials created with WebAuthn APIs cannot be utilized by U2F.

Here is the list of subtasks as I see them now, in the general order I expect to complete them (to be updated as additional things inevitably pop up):
- [x] New WebAuthn credential model
  - [x] Implement credential model
  - [x] Database migration for new credential model
  - [x] Tests for new credential model
  - [x] Database migration for legacy U2F registrations to new credential model
- [x] WebAuthn instance/handler
  - [x] Add WebAuthn handler initiation to web app startup
  - [x] Determine best way to get handler to routes (middleware?)
- [ ] WebAuthn login flow
  - [ ] New routes and handlers using ~~duo-labs/webauthn~~ e3b0c442/warp
  - [ ] New JavaScript using standard WebAuthn
- [ ] WebAuthn registration flow
  - [ ] New routes and handlers using ~~duo-labs/webauthn~~ e3b0c442/warp
  - [ ] New JavaScript using standard WebAuthn
- [ ] Update strings for WebAuthn
- [ ] Clean up legacy U2F
  - [ ] Remove legacy U2F models
  - [ ] Remove legacy U2F settings
  - [ ] Database migration to remove legacy U2F registration table
  - [ ] Remove legacy U2F API routes
  - [ ] Remove tstranex/u2f vendoring
  - [ ] Remove u2fApi Javascript library
  - [ ] Remove legacy U2F strings

#### Planned changes vs U2F implementation
* No user-configurable options. Will not use the optional icon, and relying party ID and origin will be inferred from `ROOT_URL`, instead of being defined explicitly as options. My opinion is that these are irrelevant, as the RPID and origin must match, meaning that there is no way to register a credential for a domain different than the domain you are visiting.

### CODE REVIEW

**Comment by e3b0c442** [comment]
I have a request for anyone looking at this -- if you have a fairly large user base with a large number of U2F registrations and would be willing to provide me a dump of the `u2f_registration` table, I would sincerely appreciate it. There shouldn't be anything sensitive in that table, just public keys. I just want to make sure the database migration works at a scale larger than me and the four YubiKeys I own.

Feel free to encrypt it with my [GPG public key](https://keys.openpgp.org/vks/v1/by-fingerprint/940863614B0B330750685FE8DD3FEC7C659DCEA2) if you are concerned about public exposure.

**Comment by e3b0c442** [comment]
The data model portion should be complete at this point.

There is a failing sqlite migration test that I haven't been able to nail down yet. If anybody has any clues here, I would sure appreciate a hand.

File: modules/auth/webauthn/webauthn.go [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+	"github.com/duo-labs/webauthn/webauthn"
+
+	"code.gitea.io/gitea/modules/setting"
+)
+
+//WebAuthn represents the global WebAuthn instance
+var WebAuthn *webauthn.WebAuthn
+
+//Init initializes the WebAuthn instance from the config.
+func Init() {
```
Comment (e3b0c442) [inline modules/auth/webauthn/webauthn.go:13]: I'm not 100% happy with this. I would prefer to use webauthn.New() in order to validate the settings, but this routers.GlobalInit() seems to be the best place to put this and there is no error handling there. Open to suggestions.

File: models/user.go [hunk 2]
```diff
@@ -1875,4 +1875,4 @@ func SyncExternalUsers(ctx context.Context) {
 			}
 		}
 	}
-}
+}
```
Comment (6543) [inline models/user.go]: unnesesary

**Comment by e3b0c442** [comment]
Quick status update here... while working to integrate duo-labs/webauthn, I ended up encountering a show-stopper bug. I have submitted an issue to have it addressed, but it may not be due to the possibility that making the change could affect others' implementations who are using the library.

That issue along with some other concerns I have about the design of that library (the large number of dependencies and its attachment directly to http requests in particular) is leading me to explore alternatives. I will do that while I wait for resolution on the duo-labs/webauthn showstopper.

**Comment by e3b0c442** [comment]
Another quick status update.

I ran into more issues trying to adapt the existing flows to duo-labs/webauthn which would have required backwards-incompatible changes to duo-labs/webauthn to accomplish within the macaron context. This led me to write a new WebAuthn implementation https://github.com/e3b0c442/warp that is standalone and implementation-agnostic (extensible, not tied to the `http.Request` object). I've made that transition on my local branch and will push the changes once I've done one or two more iterations on the `warp` library and gain some confidence in the stability of the API.

Ultimately I believe this will be the cleaner way to go, but willing to discuss any concerns around this direction.

**Comment by e3b0c442** [comment]
This hasn't fallen off of my radar, it's just not my day job. As mentioned previously, anybody else can feel free to pick this up and run with it, e3b0c442/warp should be stable enough at this point.

**Comment by 6543** [comment]
@e3b0c442 no problem just feed the stale bot some time ;)

**Comment by haraldrudell** [comment]
What would be really nice is Webauthn using Touch ID in ungoogled-chromium using challenge-response inside TLS which should be possible on the macOS Big Sur public beta available since 9/30/2020

https://webauthn.io/ has golang source code links. The site works, the golang code seems to need some help

Basically: everyone uses MacBook Pro, nobody uses Safari, fingerprint bio-metrics is safe

challenge response inside TLS is completely decentralized

The @Google implementation is a privacy nightmare

**Comment by e3b0c442** [comment]
2020 unfortunately was not kind to me time wise. I'm going to close this PR so that it doesn't deter anybody else that wants to from taking a crack. If I get to the point where I have something fully functional I'll open a new one.

**Comment by 6543** [comment]
-> #17957

**Pull request closed without merging**